### ISSUE TITLE
SpringDocWebFluxConfiguration throws NoClassDefFoundError: javax/servlet/ServletException

### ISSUE THREAD

**Issue by mraible** [springdoc/springdoc-openapi#1357]
**Describe the bug**

We're trying to migrate JHipster to use Springdoc instead of Springfox. 

https://github.com/jhipster/jhipster-bom/pull/556

In a Spring WebFlux application, I have the following dependencies:

```xml
<dependency>
    <groupId>org.springdoc</groupId>
    <artifactId>springdoc-openapi-webflux-core</artifactId>
</dependency>
<dependency>
    <groupId>org.springdoc</groupId>
    <artifactId>springdoc-openapi-data-rest</artifactId>
</dependency>
```

When the app starts, I get the following error:

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 
'openApiResource' defined in class path resource 
[org/springdoc/webflux/core/SpringDocWebFluxConfiguration.class]: Unsatisfied dependency expressed through 
method 'openApiResource' parameter 2; nested exception is 
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'responseBuilder' 
defined in class path resource [org/springdoc/webflux/core/SpringDocWebFluxConfiguration.class]: Bean 
instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: 
Failed to instantiate [org.springdoc.core.GenericResponseService]: Factory method 'responseBuilder' threw 
exception; nested exception is java.lang.NoClassDefFoundError: javax/servlet/ServletException
```

Another strange thing that happens with these two dependencies. A new webflux + MongoDB app fails with the following when running `mvn verify`:

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate 
[org.springdoc.core.GenericResponseService]: Factory method 'responseBuilder' threw exception; nested 
exception is java.lang.NoClassDefFoundError: Could not initialize class 
org.springdoc.core.GenericResponseService
```

However, if I change replace both dependencies with `springdoc-openapi-webflux-ui`, all tests pass.

**To Reproduce**

- What version of spring-boot you are using? 2.5.7
- What modules and versions of springdoc-openapi are you using? 1.5.13
- What is the actual and the expected result using OpenAPI Description (yml or json)?
- Provide with a sample code (HelloController) or Test that reproduces the problem

I can provide steps to reproduce, but it's kinda complicated. Here's an attempt:

```
git clone -b springdocs git@github.com:mshima/jhipster-bom.git
cd jhipster-bom
./mvnw install -Dgpg.skip=true
cd ..
git clone -b skip_ci-doc git@github.com:mshima/generator-jhipster.git
cd generator-jhipster
npm link
```

Then, create a new JHipster project (for example, in your ~/Downloads directory):

```
mkdir webflux-mongodb
cd webflux-mongodb
```

Create a `.yo-rc.json` file in this directory with the following contents:

```json
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "reactive": true,
    "baseName": "sampleWebfluxMongodb",
    "packageName": "tech.jhipster.sample",
    "packageFolder": "tech/jhipster/sample",
    "authenticationType": "jwt",
    "cacheProvider": "no",
    "enableHibernateCache": false,
    "websocket": false,
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": false,
    "buildTool": "maven",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en", "fr"],
    "testFrameworks": ["gatling", "cypress"],
    "serverPort": "8080",
    "jhiPrefix": "jhi",
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "creationTimestamp": 1596513172471,
    "jwtSecretKey": "ZjY4MTM4YjI5YzMwZjhjYjI2OTNkNTRjMWQ5Y2Q0Y2YwOWNmZTE2NzRmYzU3NTMwM2NjOTE3MTllOTM3MWRkMzcyYTljMjVmNmQ0Y2MxOTUzODc0MDhhMTlkMDIxMzI2YzQzZDM2ZDE3MmQ3NjVkODk3OTVmYzljYTQyZDNmMTQ=",
    "jhipsterVersion": "7.4.0",
    "skipClient": false,
    "skipServer": false,
    "skipUserManagement": false,
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "blueprints": [],
    "otherModules": [],
    "pages": [],
    "devServerPort": 4200,
    "clientTheme": "none",
    "clientThemeVariant": "",
    "withAdminUi": true,
    "serviceDiscoveryType": false,
    "enableGradleEnterprise": false
  }
}
```

Then, run `jhipster`. 

Once the app is finished, run `./mvnw verify` to see the test failures.

To start the app:

```
docker-compose -f src/main/docker/mongodb.yml up -d
./mvnw
```

**Comment by mshima** [comment]
Unfortunately, webflux + data-rest still not fixed.
We are having the following error related to hateoas now:
```
Error:  tech.jhipster.sample.service.MailServiceIT.testSendActivationEmail  Time elapsed: 0.001 s  <<< ERROR!
1105
java.lang.IllegalStateException: Failed to load ApplicationContext
1106
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'modelConverterRegistrar' defined in class path resource [org/springdoc/core/SpringDocConfiguration.class]: Unsatisfied dependency expressed through method 'modelConverterRegistrar' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'collectionModelContentConverter' defined in class path resource [org/springdoc/hateoas/SpringDocHateoasConfiguration.class]: Unsatisfied dependency expressed through method 'collectionModelContentConverter' parameter 1; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.hateoas.server.LinkRelationProvider' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
1107
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'collectionModelContentConverter' defined in class path resource [org/springdoc/hateoas/SpringDocHateoasConfiguration.class]: Unsatisfied dependency expressed through method 'collectionModelContentConverter' parameter 1; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.hateoas.server.LinkRelationProvider' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
1108
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.hateoas.server.LinkRelationProvider' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
1109
```

**Comment by bnasslahsen** [comment]
@mraible,

**Comment by mshima** [comment]
I don't have the time to go further now.

**Comment by mshima** [comment]
Thanks @bnasslahsen

**Comment by bnasslahsen** [comment]
@mshima, 

I have reviewed the cause of your error.
This is mainly related to the dependency `springdoc-openapi-data-rest` which is only relevant if the project uses `spring-boot-starter-data-rest`. 
As you only need `spring-webflux` support, `springdoc-openapi-webflux-core` dependency should be enough.

Once you remove the dependency `springdoc-openapi-data-rest`, you will get another error:
`Caused by: java.io.FileNotFoundException: class path resource [org/springdoc/webmvc/core/MultipleOpenApiSupportConfiguration.class] cannot be opened because it does not exist`.
For this error, you will have to adapt the  `tech.jhipster.config.apidoc.OpenApiAutoConfiguration`, which references this class a wrong class for webflux.

This is a sample configuration, that made it work:
```java
package tech.jhipster.config.apidoc;

import io.swagger.v3.oas.models.OpenAPI;
import org.springdoc.core.converters.models.Pageable;
import tech.jhipster.config.JHipsterProperties;

import org.springframework.boot.autoconfigure.AutoConfigureAfter;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import static org.springdoc.core.SpringDocUtils.getConfig;
import static tech.jhipster.config.JHipsterConstants.SPRING_PROFILE_API_DOCS;

/**
 * OpenApi OpenAPI configuration.
 * <p>
 * Warning! When having a lot of REST endpoints, OpenApi can become a performance issue.
 * In that case, you can use the "no-api-docs" Spring profile, so that this bean is ignored.
 */
@Configuration
@ConditionalOnWebApplication
@ConditionalOnClass({
		OpenAPI.class,
})
@Profile(SPRING_PROFILE_API_DOCS)
@AutoConfigureAfter(JHipsterProperties.class)
public class OpenApiAutoConfiguration {

	static {
		getConfig().replaceParameterObjectWithClass(org.springframework.data.domain.Pageable.class, Pageable.class);
	}
}
```

If you don't want the `sprindoc-openapi` beans to be loaded depending on the profile: `api-docs`, you can add at the end, of the generated `application.yml`: (So that the class `OpenApiAutoConfiguration` can be generic, not depending on any `springdoc-openapi` configuration classes)

```yaml
---
spring:
  config:
    activate:
      on-profile: '!api-docs'
springdoc:
    api-docs:
      enabled: false
```

I will keep this ticket open until your migration is complete.
If you are facing any other issues, don't hesitate to share them here.

**Comment by mshima** [comment]
@bnasslahsen thanks for the detailed review

> I have reviewed the cause of your error.
> This is mainly related to the dependency `springdoc-openapi-data-rest` which is only relevant if the project uses `spring-boot-starter-data-rest`.

I didn't realized that.
https://springdoc.org point to it for Pageable support. So I supposed it was necessary and included it.
Even with everything in place that I've removed https://github.com/jhipster/jhipster-bom/pull/556/commits/40a416c5a21f23e72578287f78e4f7ae34d4d85c and now reverted to drop data-rest.

> If you don't want the `sprindoc-openapi` beans to be loaded depending on the profile: `api-docs`, you can add at the end, of the generated `application.yml`: (So that the class `OpenApiAutoConfiguration` can be generic, not depending on any `springdoc-openapi` configuration classes)

Thanks for that. I will use it for sure.

**Comment by mshima** [comment]
@bnasslahsen if you still have the project, I've experienced https://github.com/springdoc/springdoc-openapi/issues/1270.
So cache is disabled for now.

**Comment by mshima** [comment]
> @bnasslahsen if you still have the project, I've experienced #1270.
> So cache is disabled for now.

Tracked down to a single OpenAPIService been used for every group.

**Comment by bnasslahsen** [comment]
@mshima, 

Exactly, you are right! When using groups, the `OpenAPIService` scope should be `PROTOTYPE`.
This a sample code how it's implemented in `springdoc-openapi`:

https://github.com/springdoc/springdoc-openapi/blob/cfe86b6554bd46e2e4b74f686d1e9837fee65f29/springdoc-openapi-common/src/main/java/org/springdoc/core/SpringdocBeanFactoryConfigurer.java#L86-L92

This is not working in your case, as you have declared in `JHipsterSpringDocAutoConfiguration` a conflicting condition:
- `@AutoConfigureBefore(SpringDocConfiguration.class)`  vs `@Import(SpringDocConfiguration.class)` . This causes the `CacheOrGroupedOpenApiCondition` not to be verified.

I believe the **straight forward solution** will be as follow (The less of `springdoc-openapi` Beans overriding/duplication):

```java
package tech.jhipster.config.apidoc;

import io.swagger.v3.oas.models.OpenAPI;
import org.springdoc.core.SpringDocConfiguration;
import tech.jhipster.config.JHipsterProperties;

import org.springframework.boot.autoconfigure.AutoConfigureAfter;
import org.springframework.boot.autoconfigure.AutoConfigureBefore;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.context.annotation.Profile;

import static tech.jhipster.config.JHipsterConstants.SPRING_PROFILE_API_DOCS;

/**
 * OpenApi OpenAPI configuration.
 * <p>
 * Warning! When having a lot of REST endpoints, OpenApi can become a performance issue.
 * In that case, you can use the "no-api-docs" Spring profile, so that this bean is ignored.
 */
@Configuration
@ConditionalOnWebApplication
@ConditionalOnClass(OpenAPI.class)
@Profile(SPRING_PROFILE_API_DOCS)
@AutoConfigureBefore(SpringDocConfiguration.class)
@AutoConfigureAfter(JHipsterProperties.class)
@Import(JHipsterSpringDocGroupsConfiguration.class)
public class JHipsterSpringDocAutoConfiguration {}
```
As you can see, i have also not loaded JHipsterSpringDocWebFluxConfiguration or JHipsterSpringDocMvcConfiguration as the loading of `MultipleOpenApiWebMvcResource` and `MultipleOpenApiWebFluxResource` is already done by the `sprindoc-openapi` starter.

Of course, you can also remove `springdoc.cache.disabled = true`, from the generated `application.yml`.

Don't hesitate to let me know, if these instructions help you solve the cache issue.

**Comment by mshima** [comment]
@bnasslahsen just an update.
Your solution worked, fixed the problem.
Is there is a way to detect this conflict?

- Maybe the pageable config could be moved to a config in common package, so we can import it:
https://github.com/springdoc/springdoc-openapi/blob/6c86b61cd666e7de1cee926ac34c47243f6ff741/springdoc-openapi-data-rest/src/main/java/org/springdoc/data/rest/SpringDocDataRestConfiguration.java#L88-L104

- There is a conditional at hateos package:
https://github.com/springdoc/springdoc-openapi/blob/6c86b61cd666e7de1cee926ac34c47243f6ff741/springdoc-openapi-hateoas/src/main/java/org/springdoc/hateoas/SpringDocHateoasConfiguration.java#L55
but not at data-rest package:
https://github.com/springdoc/springdoc-openapi/blob/6c86b61cd666e7de1cee926ac34c47243f6ff741/springdoc-openapi-data-rest/src/main/java/org/springdoc/data/rest/SpringDocDataRestConfiguration.java#L127-L131

I really appreciate your help.
I think this bug can be closed now.

**Comment by bnasslahsen** [comment]
@mshima, 

I agree with your enhancement proposal.
This would make sense, to make it default behaviour (out of the box - without extra declaration in the applications level)
I would also add a property, to disable this default behavior. (in case some projects wants to keep the previous mode)
